From ae0facc9d2c656dee517e800c92bb0c55dca2601 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hmehrtens@maxlinear.com>
Date: Thu, 1 Apr 2021 14:59:06 +0200
Subject: [PATCH] PONRTSYS-8780: Revert "PONRTSYS-8435: Fix TC Queue mapping
 cleanup"

This reverts commit 5caaf6908897c28176aac6a47f5064934dd2a1d4.

This breaks some BBF test cases.
---
 drivers/net/datapath/pon_qos/pon_qos_tc_qmap.c | 85 ++++++--------------------
 drivers/net/datapath/pon_qos/pon_qos_tc_qmap.h |  2 -
 drivers/net/datapath/pon_qos/pon_qos_tc_qos.c  | 17 ++----
 3 files changed, 26 insertions(+), 78 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.c b/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.c
index 0ff55775a5ac..53f675b9023b 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0
 /******************************************************************************
  *
- * Copyright (c) 2020 - 2021 MaxLinear, Inc.
+ * Copyright (c) 2020 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
@@ -23,9 +23,7 @@ struct flower_cls_map {
 	struct flow_dissector_key_vlan key;
 	struct flow_dissector_key_vlan mask;
 	struct net_device *dev;
-	int in_ifi;
-	int tc;
-	bool mapped;
+	struct net_device *indev;
 	struct list_head list;
 };
 
@@ -125,38 +123,11 @@ static int pon_qos_parse_tc_flower(struct tc_cls_flower_offload *f,
 	return ret;
 }
 
-static int __pon_qos_tc_map(struct net_device *dev,
-			    struct flower_cls_map *map,
-			    struct net_device *indev,
-			    int tc, bool en)
-{
-	struct pon_qos_qmap_tc qmap = { 0 };
-	int ret;
-
-	qmap.indev = indev;
-	if (!indev)
-		qmap.subif = tc;
-	else
-		qmap.subif = map->key.vlan_priority;
-	/* when Q is removed we should unmap the same classid */
-	qmap.handle = map->classid;
-	qmap.tc_cookie = map->tc_cookie;
-	ret = pon_qos_update_qmap(dev, &qmap, en);
-	if (ret < 0) {
-		netdev_err(dev, "%s: queue unmap fail\n", __func__);
-		return -EINVAL;
-	}
-
-	map->tc = qmap.subif;
-	map->mapped = en;
-
-	return 0;
-}
-
 int pon_qos_tc_map(struct net_device *dev, struct tc_cls_flower_offload *f)
 {
 	struct flower_cls_map *map = NULL;
 	struct net_device *indev = NULL;
+	struct pon_qos_qmap_tc qmap = { 0 };
 	char tc_cookie = PON_QOS_TC_COOKIE_EMPTY;
 	int ifi = (int)*(int *)f->key;
 	int ret = 0;
@@ -176,9 +147,13 @@ int pon_qos_tc_map(struct net_device *dev, struct tc_cls_flower_offload *f)
 		goto err_put_indev;
 
 	map->dev = dev;
-	map->in_ifi = ifi;
+	map->indev = indev;
+	qmap.indev = map->indev;
+	qmap.subif = map->key.vlan_priority;
+	qmap.handle = map->classid;
+	qmap.tc_cookie = tc_cookie;
 
-	ret = __pon_qos_tc_map(dev, map, indev, 0, true);
+	ret = pon_qos_update_qmap(dev, &qmap, 1);
 	if (ret < 0) {
 		netdev_err(dev, "%s: queue map fail\n", __func__);
 		goto err;
@@ -192,17 +167,13 @@ int pon_qos_tc_map(struct net_device *dev, struct tc_cls_flower_offload *f)
 		goto err;
 
 	list_add(&map->list, &tc_class_list);
-	if (indev)
-		dev_put(indev);
 
 	return 0;
-
 err:
 	kfree(map);
 err_put_indev:
 	if (indev)
 		dev_put(indev);
-
 	return ret;
 }
 EXPORT_SYMBOL(pon_qos_tc_map);
@@ -210,6 +181,7 @@ EXPORT_SYMBOL(pon_qos_tc_map);
 int pon_qos_tc_unmap(struct net_device *dev, void *list_node)
 {
 	struct flower_cls_map *map;
+	struct pon_qos_qmap_tc qmap = { 0 };
 	int ret = 0;
 
 	if (!list_node)
@@ -217,39 +189,22 @@ int pon_qos_tc_unmap(struct net_device *dev, void *list_node)
 
 	map = list_node;
 
-	if (map->mapped) {
-		ret = __pon_qos_tc_map(dev, map, NULL, map->tc, false);
-		if (ret < 0) {
-			netdev_err(dev, "%s: queue unmap fail\n", __func__);
-			return ret;
-		}
+	qmap.indev = map->indev;
+	qmap.subif = map->key.vlan_priority;
+	qmap.handle = map->classid;
+	qmap.tc_cookie = map->tc_cookie;
+	ret = pon_qos_update_qmap(dev, &qmap, 0);
+	if (ret < 0) {
+		netdev_err(dev, "%s: queue unmap fail\n", __func__);
+		ret = -EINVAL;
 	}
 
 	list_del(&map->list);
+	if (map->indev)
+		dev_put(map->indev);
 	kfree(map);
 
 	return 0;
 }
 EXPORT_SYMBOL(pon_qos_tc_unmap);
 
-int pon_qos_tc_classid_unmap(u32 classid)
-{
-	struct flower_cls_map *p, *n;
-	int ret;
-
-	list_for_each_entry_safe(p, n, &tc_class_list, list) {
-		if (p->classid != classid)
-			continue;
-
-		if (!p->mapped)
-			continue;
-
-		ret = __pon_qos_tc_map(p->dev, p, NULL, p->tc, false);
-		if (ret < 0) {
-			netdev_err(p->dev, "%s: queue unmap fail\n", __func__);
-			return ret;
-		}
-	}
-
-	return 0;
-}
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.h b/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.h
index 716089afee82..72586ae49edc 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.h
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_qmap.h
@@ -1,7 +1,6 @@
 /* SPDX-License-Identifier: GPL-2.0 */
 /******************************************************************************
  *
- * Copyright (c) 2021 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
@@ -10,6 +9,5 @@
 
 int pon_qos_tc_map(struct net_device *dev, struct tc_cls_flower_offload *f);
 int pon_qos_tc_unmap(struct net_device *dev, void *list_node);
-int pon_qos_tc_classid_unmap(u32 classid);
 
 #endif
diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
index 1992f35ad032..2147e5a38f7a 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0
 /******************************************************************************
  *
- * Copyright (c) 2020 - 2021 MaxLinear, Inc.
+ * Copyright (c) 2020 MaxLinear, Inc.
  * Copyright (c) 2020 Intel Corporation
  *
  *****************************************************************************/
@@ -10,7 +10,6 @@
 #include <net/datapath_api_qos.h>
 #include <net/pon_qos_tc.h>
 #include "pon_qos_tc_qos.h"
-#include "pon_qos_tc_qmap.h"
 
 static LIST_HEAD(port_list);
 
@@ -523,7 +522,7 @@ static int pon_qos_get_subif_idx(struct net_device *dev, int *subif)
 		return -ENODEV;
 	}
 
-	*subif = dp_subif.subif_groupid;
+	*subif = dp_subif.subif;
 
 	return 0;
 }
@@ -962,13 +961,13 @@ int pon_qos_update_qmap(struct net_device *dev,
 		}
 	}
 
-	if (q_tc->indev && en) {
-		ret = pon_qos_get_subif_idx(q_tc->indev,
-					    &q_tc->subif);
+	if (q_tc->indev) {
+		ret = pon_qos_get_subif_idx(q_tc->indev, &tc);
 		if (ret < 0)
 			return -ENODEV;
+	} else {
+		tc = q_tc->subif;
 	}
-	tc = q_tc->subif;
 
 	netdev_dbg(dev, "port: %d deq_i: %d qid: %d handle: %#x tc: %d ds:%s\n",
 		   sch->port, sch->deq_idx, qid->qid, q_tc->handle, tc,
@@ -1128,10 +1127,6 @@ static int __pon_qos_qdata_remove(struct net_device *dev,
 	netdev_dbg(dev, "%s: pid:%#x class/handle:%#x\n",
 		   __func__, p->parent, p->handle);
 
-	ret = pon_qos_tc_classid_unmap(p->handle);
-	if (ret)
-		return ret;
-
 	/* Remove all successors before removing ancestor */
 	ret = pon_qos_qdata_child_remove(dev, qid, p->handle);
 	if (ret)
